saleor/cart/decorators.py

```diff
@@ -11,54 +11,69 @@ def get_user_open_cart_token(user):
         return cart.token
 
 
-def find_and_assign_cart(request, response):
-    """Assign cart from cookie to request user"""
+def find_and_assign_cart(request):
+    """Assign cart from cookie to request user and close """
     cookie = request.get_signed_cookie(Cart.COOKIE_NAME, default=None)
     if cookie:
-        cart = Cart.objects.open().filter(token=cookie).first()
-        if cart is not None and cart.user is None:
+        cart = Cart.objects.open().filter(token=cookie, user=None).first()
+        if cart is not None:
             cart.change_user(request.user)
-        response.delete_cookie(Cart.COOKIE_NAME)
+            carts_to_close = Cart.objects.open().filter(user=request.user)
+            carts_to_close = carts_to_close.exclude(token=cookie)
+            for to_close in carts_to_close:
+                to_close.change_status(Cart.CANCELED)
```

Comment:
I'd recommend to use `update()` method instead of executing couple of queries in a loop. You can update `last_updated` field here as well.